File: app/src/main/java/com/espressif/ui/BLEProvisionLanding.java
// Copyright 2018 Espressif Systems (Shanghai) PTE LTD
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
package com.espressif.ui;

import android.Manifest;
import android.bluetooth.BluetoothAdapter;
import android.bluetooth.BluetoothDevice;
import android.bluetooth.BluetoothManager;
import android.content.Context;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.graphics.Color;
import android.os.Bundle;
import android.os.Vibrator;
import android.support.v4.app.ActivityCompat;
import android.support.v4.content.ContextCompat;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.Toolbar;
import android.util.Log;
import android.view.HapticFeedbackConstants;
import android.view.View;
import android.widget.AdapterView;
import android.widget.ArrayAdapter;
import android.widget.Button;
import android.widget.ListView;
import android.widget.ProgressBar;
import android.widget.TextView;
import android.widget.Toast;

import com.espressif.avs.ConfigureAVS;
import com.espressif.provision.Provision;
import com.espressif.provision.R;
import com.espressif.provision.security.Security;
import com.espressif.provision.security.Security0;
import com.espressif.provision.security.Security1;
import com.espressif.provision.session.Session;
import com.espressif.provision.transport.BLETransport;
import com.espressif.provision.transport.ResponseListener;
import com.google.protobuf.InvalidProtocolBufferException;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.UUID;

import avs.Avsconfig;

import static com.espressif.avs.ConfigureAVS.AVS_CONFIG_PATH;

public class BLEProvisionLanding extends AppCompatActivity {

    private static final String TAG = "Espressif::" + BLEProvisionLanding.class.getSimpleName();

    private static final String PROOF_OF_POSSESSION = "abcd1234";

    private Button btnScan;
    private ProgressBar progressBar;
    private ListView listView;

    private BluetoothAdapter bleAdapter;
    private ArrayList<BluetoothDevice> bluetoothDevices;

    private Session session;
    private Security security;
    public static BLETransport bleTransport;
    private BLETransport.BLETransportListener transportListener;
    // FIXME : Remove static BLE_TRANSPORT and think for another solution.

    private String configUUID;
    private String avsConfigUUID;
    private String serviceUUID;
    private String sessionUUID;
    private String deviceName;
    private String deviceNamePrefix;
    private boolean isScanning;
    private boolean isDeviceConnected;
    private ArrayAdapter<String> adapter;

    @Override
    protected void onCreate(Bundle savedInstanceState) {

        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_bleprovision_landing);
        Toolbar toolbar = (Toolbar) findViewById(R.id.toolbar);
        toolbar.setTitle(R.string.connect_to_device_title);
        setSupportActionBar(toolbar);

        serviceUUID = getIntent().getStringExtra(BLETransport.SERVICE_UUID_KEY);
        sessionUUID = getIntent().getStringExtra(BLETransport.SESSION_UUID_KEY);
        configUUID = getIntent().getStringExtra(BLETransport.CONFIG_UUID_KEY);
        avsConfigUUID = getIntent().getStringExtra(ConfigureAVS.AVS_CONFIG_UUID_KEY);
        deviceNamePrefix = getIntent().getStringExtra(BLETransport.DEVICE_NAME_PREFIX_KEY);

        btnScan = findViewById(R.id.btn_scan);
        listView = findViewById(R.id.ble_devices_list);
        progressBar = findViewById(R.id.ble_landing_progress_indicator);

        isDeviceConnected = false;
        bluetoothDevices = new ArrayList<>();
        ArrayList<String> bleNames = new ArrayList<>();

        adapter = new ArrayAdapter<>(this,
                android.R.layout.simple_list_item_1,
                android.R.id.text1,
                bleNames);

        // Assign adapter to ListView
        listView.setAdapter(adapter);
        listView.setOnItemClickListener(new AdapterView.OnItemClickListener() {
            @Override
            public void onItemClick(AdapterView<?> adapterView, View view, int pos, long l) {
                isDeviceConnected = false;
                btnScan.setVisibility(View.GONE);
                listView.setVisibility(View.GONE);
                progressBar.setVisibility(View.VISIBLE);
                deviceName = bluetoothDevices.get(pos).getName();
                bleTransport.connect(bluetoothDevices.get(pos));
            }
        });

        btnScan.setOnClickListener(new View.OnClickListener() {

            @Override
            public void onClick(View view) {

                Vibrator vib = (Vibrator) getSystemService(VIBRATOR_SERVICE);
                vib.vibrate(HapticFeedbackConstants.VIRTUAL_KEY);

                startScan();
            }
        });

        transportListener = new BLETransport.BLETransportListener() {

            @Override
            public void onPeripheralsFound(ArrayList<BluetoothDevice> devices) {

                boolean deviceExists = false;
                for (BluetoothDevice device : devices) {
                    for (BluetoothDevice alreadyHere : bluetoothDevices) {
                        if (device.equals(alreadyHere)) {
                            deviceExists = true;
                            break;
                        }
                    }
                    if (!deviceExists) {
                        bluetoothDevices.add(device);
                        adapter.add(device.getName());
                    }
                    deviceExists = false;
                }
                stopScan();
            }

            @Override
            public void onPeripheralsNotFound() {

                runOnUiThread(new Runnable() {
                    @Override
                    public void run() {
                        Toast.makeText(BLEProvisionLanding.this,
                                "No Bluetooth devices found!",
                                Toast.LENGTH_LONG)
                                .show();
                    }
                });
                stopScan();
            }

            @Override
            public void onPeripheralConfigured(BluetoothDevice device) {

                bleDeviceConfigured(true);
            }

            @Override
            public void onPeripheralNotConfigured(BluetoothDevice device) {

                isScanning = false;
                btnScan.setEnabled(true);
                btnScan.setAlpha(1f);
                progressBar.setVisibility(View.GONE);
                listView.setVisibility(View.VISIBLE);
                bleDeviceConfigured(false);
            }

            @Override
            public void onPeripheralDisconnected(Exception e) {
                runOnUiThread(new Runnable() {
                    @Override
                    public void run() {
                        Toast.makeText(BLEProvisionLanding.this,
                                "Bluetooth device disconnected.",
                                Toast.LENGTH_LONG)
                                .show();
                        // TODO check for stopScan();
                    }
                });
            }

            @Override
            public void onFailure(final Exception e) {
                runOnUiThread(new Runnable() {
                    @Override
                    public void run() {
                        Toast.makeText(BLEProvisionLanding.this,
                                "Bluetooth connection failed : " + e.getMessage(),
                                Toast.LENGTH_LONG)
                                .show();
                    }
                });
            }
        };

        HashMap<String, String> configUUIDMap = new HashMap<>();
        configUUIDMap.put(Provision.PROVISIONING_CONFIG_PATH, configUUID);
        configUUIDMap.put("prov-scan", "0000ff50-0000-1000-8000-00805f9b34fb");
        if (avsConfigUUID != null) {
            configUUIDMap.put(ConfigureAVS.AVS_CONFIG_PATH, avsConfigUUID);
        }

        bleTransport = new BLETransport(this,
                UUID.fromString(serviceUUID),
                UUID.fromString(sessionUUID),
                configUUIDMap,
                deviceNamePrefix,
                3000);

        boolean isBLEEnabled = this.checkBLEPermissions();
        if (!isBLEEnabled) {
            btnScan.setEnabled(false);
            btnScan.setAlpha(0.5f);
            btnScan.setTextColor(Color.WHITE);
            requestBluetoothEnable();
        } else {
            btnScan.setEnabled(true);
            btnScan.setAlpha(1f);
            btnScan.setTextColor(Color.WHITE);
            startScan();
        }
    }

    @Override
    protected void onResume() {
        super.onResume();

        Log.d(TAG, "ON RESUME");
        if (isDeviceConnected) {
            bleTransport.disconnect();
            final TextView bleInstructions = findViewById(R.id.bluetooth_status_message);
            bleInstructions.setText(R.string.enable_bluetooth_instructions);
        }
    }

    private void bleDeviceConfigured(final Boolean isConfigured) {

        runOnUiThread(new Runnable() {
            @Override
            public void run() {

                if (isConfigured) {

                    isDeviceConnected = true;
//                    finish();
//                    goToProofOfPossessionActivity();
                    initSession();

                } else {
                    Toast.makeText(BLEProvisionLanding.this,
                            "Bluetooth device could not be configured. Please try another device.",
                            Toast.LENGTH_LONG)
                            .show();
                }
            }
        });
    }

    private void startScan() {

        isScanning = true;
        bluetoothDevices.clear();
        adapter.clear();
        updateProgressAndScanBtn();
        bleTransport.scan(transportListener);
    }

    private void stopScan() {

        isScanning = false;
        updateProgressAndScanBtn();
    }

    /**
     * This method will update UI (Scan button enable / disable and progressbar visibility)
     */
    private void updateProgressAndScanBtn() {

        if (isScanning) {

            btnScan.setEnabled(false);
            btnScan.setAlpha(0.5f);
            btnScan.setTextColor(Color.WHITE);
            progressBar.setVisibility(View.VISIBLE);
            listView.setVisibility(View.GONE);

        } else {

            btnScan.setEnabled(true);
            btnScan.setAlpha(1f);
            progressBar.setVisibility(View.GONE);
            listView.setVisibility(View.VISIBLE);
        }
    }

    private void goToProofOfPossessionActivity() {

        Intent alexaProvisioningIntent = new Intent(getApplicationContext(), ProofOfPossessionActivity.class);
        alexaProvisioningIntent.putExtras(getIntent());
        alexaProvisioningIntent.putExtra(LoginWithAmazon.KEY_IS_PROVISIONING, true);
        startActivity(alexaProvisioningIntent);
    }

    private boolean checkBLEPermissions() {
        boolean isBLEEnabled = false;
        final BluetoothManager bluetoothManager = (BluetoothManager) getSystemService(Context.BLUETOOTH_SERVICE);
        bleAdapter = bluetoothManager.getAdapter();
        if (bleAdapter == null || !bleAdapter.isEnabled()) {
            isBLEEnabled = false;
        } else if (!hasLocationPermissions()) {
            requestLocationPermission();
        } else {
            isBLEEnabled = true;
        }

        return isBLEEnabled;
    }

    private void requestBluetoothEnable() {
        Intent enableBtIntent = new Intent(BluetoothAdapter.ACTION_REQUEST_ENABLE);
        startActivityForResult(enableBtIntent, Provision.REQUEST_ENABLE_BLE_CODE);
    }

    private boolean hasLocationPermissions() {
        return ContextCompat.checkSelfPermission(this,
                Manifest.permission.ACCESS_FINE_LOCATION) == PackageManager.PERMISSION_GRANTED;
    }

    private void requestLocationPermission() {
        ActivityCompat.requestPermissions(this,
                new String[]{Manifest.permission.ACCESS_FINE_LOCATION}, 102);
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, Intent data) {

        Log.e(TAG, "On Activity Result : requestCode : " + requestCode + ", resultCode : " + resultCode);

        if (requestCode == Provision.REQUEST_PROVISIONING_CODE && resultCode == RESULT_OK) {
            setResult(resultCode);
            finish();
        } else if (requestCode == Provision.REQUEST_ENABLE_BLE_CODE && resultCode == RESULT_OK) {
            boolean isEnable = checkBLEPermissions();

            if (isEnable) {

                HashMap<String, String> configUUIDMap = new HashMap<>();
                configUUIDMap.put(Provision.PROVISIONING_CONFIG_PATH, configUUID);
                configUUIDMap.put("prov-scan", "0000ff50-0000-1000-8000-00805f9b34fb");
                if (avsConfigUUID != null) {
                    configUUIDMap.put(ConfigureAVS.AVS_CONFIG_PATH, avsConfigUUID);
                }
                startScan();
            } else {
                Log.e(TAG, "Error");
            }
        }
    }

    @Override
    public void onRequestPermissionsResult(int requestCode,
                                           String permissions[], int[] grantResults) {
        switch (requestCode) {
            case Provision.REQUEST_PERMISSIONS_CODE: {
                // If request is cancelled, the result arrays are empty.
                if (grantResults.length > 0
                        && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
                    checkBLEPermissions();
                }
            }
            break;
        }
    }

    private void getStatus() {

        Avsconfig.CmdSignInStatus configRequest = Avsconfig.CmdSignInStatus.newBuilder()
                .setDummy(123)
                .build();
        Avsconfig.AVSConfigMsgType msgType = Avsconfig.AVSConfigMsgType.TypeCmdSignInStatus;
        Avsconfig.AVSConfigPayload payload = Avsconfig.AVSConfigPayload.newBuilder()
                .setMsg(msgType)
                .setCmdSigninStatus(configRequest)
                .build();

        byte[] message = security.encrypt(payload.toByteArray());

        BLEProvisionLanding.bleTransport.sendConfigData(AVS_CONFIG_PATH, message, new ResponseListener() {

            @Override
            public void onSuccess(byte[] returnData) {

                Avsconfig.AVSConfigStatus deviceStatus = processSignInStatusResponse(returnData);
                Log.d(TAG, "SignIn Status Received : " + deviceStatus);
                finish();

                if (deviceStatus.equals(Avsconfig.AVSConfigStatus.SignedIn)) {

                    goToProvisionActivity();

                } else {
                    goToLoginActivity();
                }
            }

            @Override
            public void onFailure(Exception e) {
                Log.d(TAG, "Error in getting status");
                e.printStackTrace();
            }
        });
    }

    private Avsconfig.AVSConfigStatus processSignInStatusResponse(byte[] responseData) {

        Avsconfig.AVSConfigStatus status = Avsconfig.AVSConfigStatus.InvalidState;
        byte[] decryptedData = this.security.decrypt(responseData);

        try {

            Avsconfig.AVSConfigPayload payload = Avsconfig.AVSConfigPayload.parseFrom(decryptedData);
            Avsconfig.RespSignInStatus signInStatus = payload.getRespSigninStatus();
            status = signInStatus.getStatus();
            Log.d(TAG, "SignIn Status message " + status.getNumber() + status.toString());

        } catch (InvalidProtocolBufferException e) {
            e.printStackTrace();
        }
        return status;
    }

    private void goToLoginActivity() {

        Intent alexaProvisioningIntent = new Intent(getApplicationContext(), LoginWithAmazon.class);
        alexaProvisioningIntent.putExtras(getIntent());
        alexaProvisioningIntent.putExtra(LoginWithAmazon.KEY_DEVICE_NAME, deviceName);
        alexaProvisioningIntent.putExtra(LoginWithAmazon.KEY_IS_PROVISIONING, true);
        alexaProvisioningIntent.putExtra(ProofOfPossessionActivity.KEY_PROOF_OF_POSSESSION, PROOF_OF_POSSESSION);
        startActivity(alexaProvisioningIntent);
    }

    private void goToProvisionActivity() {

        Intent launchProvisionInstructions = new Intent(getApplicationContext(), ProvisionActivity.class);
        launchProvisionInstructions.putExtras(getIntent());
        launchProvisionInstructions.putExtra(LoginWithAmazon.KEY_IS_PROVISIONING, true);
        launchProvisionInstructions.putExtra(ProofOfPossessionActivity.KEY_PROOF_OF_POSSESSION, PROOF_OF_POSSESSION);
        startActivityForResult(launchProvisionInstructions, Provision.REQUEST_PROVISIONING_CODE);
    }

    private void initSession() {

        final String pop = PROOF_OF_POSSESSION;
        Log.e(TAG, "POP : " + pop);
        final String securityVersion = getIntent().getStringExtra(Provision.CONFIG_SECURITY_KEY);

        if (securityVersion.equals(Provision.CONFIG_SECURITY_SECURITY1)) {
            security = new Security1(pop);
        } else {
            security = new Security0();
        }

        session = new Session(BLEProvisionLanding.bleTransport, security);

        session.sessionListener = new Session.SessionListener() {

            @Override
            public void OnSessionEstablished() {
                Log.d(TAG, "Session established");
                getStatus();
            }

            @Override
            public void OnSessionEstablishFailed(Exception e) {
                Log.d(TAG, "Session failed");
            }
        };
        session.init(null);
    }
}

File: app/src/main/java/com/espressif/ui/activities/BLEProvisionLanding.kt
// Copyright 2025 Espressif Systems (Shanghai) PTE LTD
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package com.espressif.ui.activities

import android.Manifest
import android.bluetooth.BluetoothAdapter
import android.bluetooth.BluetoothDevice
import android.bluetooth.BluetoothManager
import android.bluetooth.le.ScanResult
import android.content.Intent
import android.content.pm.PackageManager
import android.graphics.Color
import android.os.Build
import android.os.Bundle
import android.os.Handler
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.widget.AdapterView.OnItemClickListener
import android.widget.EditText
import android.widget.Toast
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.content.res.AppCompatResources
import androidx.core.app.ActivityCompat
import com.espressif.AppConstants
import com.espressif.provisioning.DeviceConnectionEvent
import com.espressif.provisioning.ESPConstants
import com.espressif.provisioning.listeners.BleScanListener
import com.espressif.ui.adapters.BleDeviceListAdapter
import com.espressif.ui.models.BleDevice
import com.espressif.ui.utils.Utils
import com.espressif.wifi_provisioning.BuildConfig
import com.espressif.wifi_provisioning.R
import com.espressif.wifi_provisioning.databinding.ActivityBleprovisionLandingBinding
import org.greenrobot.eventbus.Subscribe
import org.greenrobot.eventbus.ThreadMode

class BLEProvisionLanding : ManualProvBaseActivity() {

    companion object {

        private val TAG: String = BLEProvisionLanding::class.java.simpleName

        // Request codes
        private const val REQUEST_ENABLE_BT = 1
        private const val REQUEST_FINE_LOCATION = 2

        // Time out
        private const val DEVICE_CONNECT_TIMEOUT: Long = 20000
    }

    private lateinit var binding: ActivityBleprovisionLandingBinding

    private lateinit var deviceList: ArrayList<BleDevice>

    //    public static boolean isBleWorkDone = false;

    private var adapter: BleDeviceListAdapter? = null
    private var bleAdapter: BluetoothAdapter? = null
    private var bluetoothDevices: HashMap<BluetoothDevice, String>? = null
    private var handler: Handler? = null

    private var position = -1
    private var deviceNamePrefix: String? = null
    private var isDeviceConnected = false
    private var isConnecting = false
    private var isScanning = false

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityBleprovisionLandingBinding.inflate(layoutInflater)
        setContentView(binding.root)
        
        // Use this check to determine whether BLE is supported on the device.  Then you can
        // selectively disable BLE-related features.
        if (!packageManager.hasSystemFeature(PackageManager.FEATURE_BLUETOOTH_LE)) {
            Toast.makeText(this, R.string.error_ble_not_supported, Toast.LENGTH_SHORT).show()
            finish()
        }

        val bluetoothManager = getSystemService(BLUETOOTH_SERVICE) as BluetoothManager
        bleAdapter = bluetoothManager.adapter

        // Checks if Bluetooth is supported on the device.
        if (bleAdapter == null) {
            Toast.makeText(this, R.string.error_bluetooth_not_supported, Toast.LENGTH_SHORT).show()
            finish()
            return
        }

        isConnecting = false
        isDeviceConnected = false
        handler = Handler()
        bluetoothDevices = HashMap()
        deviceList = ArrayList()
        deviceNamePrefix = sharedPreferences?.getString(
            AppConstants.KEY_BLE_DEVICE_NAME_PREFIX,
            resources.getString(R.string.ble_device_name_prefix)
        )
        initViews()
    }

    override fun onResume() {
        super.onResume()

        // Ensures Bluetooth is enabled on the device.  If Bluetooth is not currently enabled,
        // fire an intent to display a dialog asking the user to grant permission to enable it.
        if (!bleAdapter!!.isEnabled) {
            val enableBtIntent = Intent(BluetoothAdapter.ACTION_REQUEST_ENABLE)
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) {
                if (ActivityCompat.checkSelfPermission(
                        this@BLEProvisionLanding,
                        Manifest.permission.BLUETOOTH_CONNECT
                    ) == PackageManager.PERMISSION_GRANTED
                ) {
                    startActivityForResult(enableBtIntent, REQUEST_ENABLE_BT)
                } else {
                    Log.e(TAG, "BLUETOOTH_CONNECT permission is not granted.")
                    return
                }
            } else {
                startActivityForResult(enableBtIntent, REQUEST_ENABLE_BT)
            }
        } else {
            if (!isDeviceConnected && !isConnecting) {
                startScan()
            }
        }
    }

    override fun onBackPressed() {
        if (isScanning) {
            stopScan()
        }
        super.onBackPressed()
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        Log.d(TAG, "onActivityResult, requestCode : $requestCode, resultCode : $resultCode")

        // User chose not to enable Bluetooth.
        if (requestCode == REQUEST_ENABLE_BT && resultCode == RESULT_CANCELED) {
            finish()
            return
        }

        if (requestCode == REQUEST_ENABLE_BT && resultCode == RESULT_OK) {
            startScan()
        }
    }

    override fun onRequestPermissionsResult(
        requestCode: Int,
        permissions: Array<String>,
        grantResults: IntArray
    ) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults)

        when (requestCode) {
            REQUEST_FINE_LOCATION -> {
                // If request is cancelled, the result arrays are empty.
                if (grantResults.size > 0 && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
                    startScan()
                } else if (grantResults.size > 0 && grantResults[0] == PackageManager.PERMISSION_DENIED) {
                    finish()
                }
            }
        }
    }

    @Subscribe(threadMode = ThreadMode.MAIN)
    fun onEvent(event: DeviceConnectionEvent) {
        Log.d(TAG, "ON Device Prov Event RECEIVED : " + event.eventType)
        handler!!.removeCallbacks(disconnectDeviceTask)

        when (event.eventType) {
            ESPConstants.EVENT_DEVICE_CONNECTED -> {
                Log.d(TAG, "Device Connected Event Received")
                binding.layoutBleProvisioning.bleLandingProgressIndicator.visibility = View.GONE
                isConnecting = false
                isDeviceConnected = true
                setSecurityTypeFromVersionInfo()
                val isSecure = sharedPreferences!!.getBoolean(AppConstants.KEY_SECURITY_TYPE, true)
                if (isSecure) {
                    if (securityType == AppConstants.SEC_TYPE_0) {
                        Utils.displayDeviceConnectionError(
                            this,
                            getString(R.string.error_security_mismatch)
                        )
                    } else {
                        processDeviceCapabilities()
                    }
                } else {
                    if (securityType != AppConstants.SEC_TYPE_0) {
                        Utils.displayDeviceConnectionError(
                            this,
                            getString(R.string.error_security_mismatch)
                        )
                    } else {
                        processDeviceCapabilities()
                    }
                }
            }

            ESPConstants.EVENT_DEVICE_DISCONNECTED -> {
                binding.layoutBleProvisioning.bleLandingProgressIndicator.visibility = View.GONE
                isConnecting = false
                isDeviceConnected = false
                Toast.makeText(this@BLEProvisionLanding, "Device disconnected", Toast.LENGTH_LONG)
                    .show()
            }

            ESPConstants.EVENT_DEVICE_CONNECTION_FAILED -> {
                binding.layoutBleProvisioning.bleLandingProgressIndicator.visibility = View.GONE
                isConnecting = false
                isDeviceConnected = false
                Utils.displayDeviceConnectionError(
                    this,
                    getString(R.string.error_device_connect_failed)
                )
            }
        }
    }

    private fun initViews() {
        setToolbar()

        // Set visibility of Prefix layout
        if (BuildConfig.isFilteringByPrefixAllowed) {
            binding.layoutBleProvisioning.prefixLayout.visibility = View.VISIBLE
        } else {
            binding.layoutBleProvisioning.prefixLayout.visibility = View.GONE
        }
        binding.layoutBleProvisioning.prefixValue.text = deviceNamePrefix

        // Set adapter for device list
        adapter = BleDeviceListAdapter(this, R.layout.item_ble_scan, deviceList)
        binding.layoutBleProvisioning.bleDevicesList.setAdapter(adapter)
        binding.layoutBleProvisioning.bleDevicesList.setOnItemClickListener(onDeviceCLickListener)
        binding.layoutBleProvisioning.btnScan.setOnClickListener(btnScanClickListener)
        binding.layoutBleProvisioning.btnChangePrefix.setOnClickListener(
            btnPrefixChangeClickListener
        )
    }

    private fun setToolbar() {
        setSupportActionBar(binding.titleBar.toolbar)
        supportActionBar!!.setDisplayHomeAsUpEnabled(true)
        supportActionBar!!.setDisplayShowHomeEnabled(true)
        supportActionBar!!.setTitle(R.string.title_activity_connect_device)
        binding.titleBar.toolbar.navigationIcon =
            AppCompatResources.getDrawable(this, R.drawable.ic_arrow_left)
        binding.titleBar.toolbar.setNavigationOnClickListener {
            finish()
        }
    }

    private fun hasPermissions(): Boolean {
        if (bleAdapter == null || !bleAdapter!!.isEnabled) {
            requestBluetoothEnable()
            return false
        } else if (!hasLocationAndBtPermissions()) {
            requestLocationAndBtPermission()
            return false
        }
        return true
    }

    private fun requestBluetoothEnable() {
        val enableBtIntent = Intent(BluetoothAdapter.ACTION_REQUEST_ENABLE)

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) {
            if (ActivityCompat.checkSelfPermission(
                    this,
                    Manifest.permission.BLUETOOTH_CONNECT
                ) == PackageManager.PERMISSION_GRANTED
            ) {
                startActivityForResult(enableBtIntent, REQUEST_ENABLE_BT)
            } else {
                requestLocationAndBtPermission()
            }
        } else {
            startActivityForResult(enableBtIntent, REQUEST_ENABLE_BT)
        }
    }

    private fun hasLocationAndBtPermissions(): Boolean {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) {
            val permissionsGranted =
                checkSelfPermission(Manifest.permission.ACCESS_FINE_LOCATION) == PackageManager.PERMISSION_GRANTED && checkSelfPermission(
                    Manifest.permission.BLUETOOTH_SCAN
                ) == PackageManager.PERMISSION_GRANTED && checkSelfPermission(
                    Manifest.permission.BLUETOOTH_CONNECT
                ) == PackageManager.PERMISSION_GRANTED
            return permissionsGranted
        } else {
            return checkSelfPermission(Manifest.permission.ACCESS_FINE_LOCATION) == PackageManager.PERMISSION_GRANTED
        }
    }

    private fun requestLocationAndBtPermission() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) {
            requestPermissions(
                arrayOf(
                    Manifest.permission.ACCESS_FINE_LOCATION, Manifest.permission.BLUETOOTH_SCAN,
                    Manifest.permission.BLUETOOTH_CONNECT
                ), REQUEST_FINE_LOCATION
            )
        } else {
            requestPermissions(
                arrayOf(Manifest.permission.ACCESS_FINE_LOCATION),
                REQUEST_FINE_LOCATION
            )
        }
    }

    private fun processDeviceCapabilities() {
        val deviceCaps = provisionManager!!.espDevice.deviceCapabilities

        if (deviceCaps != null) {
            if (!deviceCaps.contains("no_pop") && securityType != AppConstants.SEC_TYPE_0) {
                goToPopActivity()
            } else if (deviceCaps.contains(AppConstants.CAPABILITY_WIFI_SCAN)) {
                goToWifiScanListActivity()
            } else if (deviceCaps.contains(AppConstants.CAPABILITY_THREAD_SCAN)) {
                goToThreadScanActivity(true)
            } else if (deviceCaps.contains(AppConstants.CAPABILITY_THREAD_PROV)) {
                goToThreadScanActivity(false)
            } else {
                goToWiFiConfigActivity()
            }
        } else {
            goToWiFiConfigActivity()
        }
    }

    private fun startScan() {
        if (!hasPermissions() || isScanning) {
            return
        }

        isScanning = true
        deviceList!!.clear()
        bluetoothDevices!!.clear()

        if (ActivityCompat.checkSelfPermission(
                this,
                Manifest.permission.ACCESS_FINE_LOCATION
            ) == PackageManager.PERMISSION_GRANTED
        ) {
            provisionManager!!.searchBleEspDevices(deviceNamePrefix, bleScanListener)
            updateProgressAndScanBtn()
        } else {
            Log.e(TAG, "Not able to start scan as Location permission is not granted.")
            Toast.makeText(
                this@BLEProvisionLanding,
                "Please give location permission to start BLE scan",
                Toast.LENGTH_LONG
            ).show()
        }
    }

    private fun stopScan() {
        isScanning = false

        if (ActivityCompat.checkSelfPermission(
                this,
                Manifest.permission.ACCESS_FINE_LOCATION
            ) == PackageManager.PERMISSION_GRANTED
        ) {
            provisionManager!!.stopBleScan()
            updateProgressAndScanBtn()
        } else {
            Log.e(TAG, "Not able to stop scan as Location permission is not granted.")
            Toast.makeText(
                this@BLEProvisionLanding,
                "Please give location permission to stop BLE scan",
                Toast.LENGTH_LONG
            ).show()
        }

        if (deviceList!!.size <= 0) {
            Toast.makeText(
                this@BLEProvisionLanding,
                R.string.error_no_ble_device,
                Toast.LENGTH_SHORT
            ).show()
        }
    }

    /**
     * This method will update UI (Scan button enable / disable and progressbar visibility)
     */
    private fun updateProgressAndScanBtn() {
        if (isScanning) {
            binding.layoutBleProvisioning.btnScan.isEnabled = false
            binding.layoutBleProvisioning.btnScan.alpha = 0.5f
            binding.layoutBleProvisioning.btnScan.setTextColor(Color.WHITE)
            binding.layoutBleProvisioning.bleLandingProgressIndicator.visibility = View.VISIBLE
            binding.layoutBleProvisioning.bleDevicesList.visibility = View.GONE
        } else {
            binding.layoutBleProvisioning.btnScan.isEnabled = true
            binding.layoutBleProvisioning.btnScan.alpha = 1f
            binding.layoutBleProvisioning.bleLandingProgressIndicator.visibility = View.GONE
            binding.layoutBleProvisioning.bleDevicesList.visibility = View.VISIBLE
        }
    }

    private val btnScanClickListener = View.OnClickListener {
        bluetoothDevices!!.clear()
        adapter!!.clear()
        startScan()
    }

    private val btnPrefixChangeClickListener = View.OnClickListener { askForPrefix() }

    private val bleScanListener: BleScanListener = object : BleScanListener {
        override fun scanStartFailed() {
            Toast.makeText(
                this@BLEProvisionLanding,
                "Please turn on Bluetooth to connect BLE device",
                Toast.LENGTH_SHORT
            ).show()
        }

        override fun onPeripheralFound(device: BluetoothDevice, scanResult: ScanResult) {
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) {
                if (ActivityCompat.checkSelfPermission(
                        this@BLEProvisionLanding,
                        Manifest.permission.BLUETOOTH_CONNECT
                    ) == PackageManager.PERMISSION_GRANTED
                ) {
                    Log.d(TAG, "====== onPeripheralFound ===== " + device.name)
                }
            } else {
                Log.d(TAG, "====== onPeripheralFound ===== " + device.name)
            }

            var deviceExists = false
            var serviceUuid = ""

            if (scanResult.scanRecord!!.serviceUuids != null && scanResult.scanRecord!!.serviceUuids.size > 0) {
                serviceUuid = scanResult.scanRecord!!.serviceUuids[0].toString()
            }
            Log.d(TAG, "Add service UUID : $serviceUuid")

            if (bluetoothDevices!!.containsKey(device)) {
                deviceExists = true
            }

            if (!deviceExists) {
                val bleDevice = BleDevice(scanResult.scanRecord!!.deviceName, device)
                binding.layoutBleProvisioning.bleDevicesList.visibility = View.VISIBLE
                bluetoothDevices!![device] = serviceUuid
                deviceList!!.add(bleDevice)
                adapter!!.notifyDataSetChanged()
            }
        }

        override fun scanCompleted() {
            isScanning = false
            updateProgressAndScanBtn()
        }

        override fun onFailure(e: Exception) {
            Log.e(TAG, e.message!!)
            e.printStackTrace()
        }
    }

    private val onDeviceCLickListener =
        OnItemClickListener { adapterView, view, position, l ->
            stopScan()
            isConnecting = true
            isDeviceConnected = false
            binding.layoutBleProvisioning.btnScan.visibility = View.GONE
            binding.layoutBleProvisioning.bleDevicesList.visibility = View.GONE
            binding.layoutBleProvisioning.bleLandingProgressIndicator.visibility = View.VISIBLE
            this@BLEProvisionLanding.position = position
            val bleDevice = adapter!!.getItem(position)
            val uuid = bluetoothDevices!![bleDevice!!.bluetoothDevice]
            Log.d(
                TAG,
                "=================== Connect to device : " + bleDevice.name + " UUID : " + uuid
            )
            if (ActivityCompat.checkSelfPermission(
                    this@BLEProvisionLanding,
                    Manifest.permission.ACCESS_FINE_LOCATION
                ) == PackageManager.PERMISSION_GRANTED
            ) {
                provisionManager?.espDevice!!.connectBLEDevice(bleDevice.bluetoothDevice, uuid)
                handler!!.postDelayed(disconnectDeviceTask, DEVICE_CONNECT_TIMEOUT)
            } else {
                Log.e(TAG, "Not able to connect device as Location permission is not granted.")
                Toast.makeText(
                    this@BLEProvisionLanding,
                    "Please give location permission to connect device",
                    Toast.LENGTH_LONG
                ).show()
            }
        }

    private val disconnectDeviceTask = Runnable {
        Log.e(TAG, "Disconnect device")
        binding.layoutBleProvisioning.bleLandingProgressIndicator.visibility = View.GONE
        Utils.displayDeviceConnectionError(
            this@BLEProvisionLanding,
            getString(R.string.error_device_not_supported)
        )
    }

    private fun askForPrefix() {
        val builder = AlertDialog.Builder(this)
        builder.setCancelable(true)

        val layoutInflaterAndroid = LayoutInflater.from(this)
        val view = layoutInflaterAndroid.inflate(R.layout.dialog_prefix, null)
        builder.setView(view)
        val etPrefix = view.findViewById<EditText>(R.id.et_prefix)
        etPrefix.setText(deviceNamePrefix)
        etPrefix.setSelection(etPrefix.text.length)

        // Set up the buttons
        builder.setPositiveButton(R.string.btn_save) { dialog, which ->
            var prefix = etPrefix.text.toString()
            if (prefix != null) {
                prefix = prefix.trim { it <= ' ' }
            }

            val editor = sharedPreferences!!.edit()
            editor.putString(AppConstants.KEY_BLE_DEVICE_NAME_PREFIX, prefix)
            editor.apply()
            deviceNamePrefix = prefix
            binding.layoutBleProvisioning.prefixValue.text = prefix
            startScan()
        }

        builder.setNegativeButton(R.string.btn_cancel) { dialog, which -> }
        builder.show()
    }

    private fun goToPopActivity() {
        finish()
        val popIntent = Intent(applicationContext, ProofOfPossessionActivity::class.java)
        popIntent.putExtra(AppConstants.KEY_DEVICE_NAME, deviceList!![position].name)
        startActivity(popIntent)
    }

    private fun goToWifiScanListActivity() {
        finish()
        val wifiListIntent = Intent(applicationContext, WiFiScanActivity::class.java)
        wifiListIntent.putExtra(AppConstants.KEY_DEVICE_NAME, deviceList!![position].name)
        startActivity(wifiListIntent)
    }

    private fun goToThreadScanActivity(scanCapAvailable: Boolean) {
        finish()
        val threadConfigIntent = Intent(applicationContext, ThreadConfigActivity::class.java)
        threadConfigIntent.putExtra(AppConstants.KEY_DEVICE_NAME, deviceList!![position].name)
        threadConfigIntent.putExtra(AppConstants.KEY_THREAD_SCAN_AVAILABLE, scanCapAvailable)
        startActivity(threadConfigIntent)
    }

    private fun goToWiFiConfigActivity() {
        finish()
        val wifiConfigIntent = Intent(applicationContext, WiFiConfigActivity::class.java)
        startActivity(wifiConfigIntent)
    }
}
